crates/red_knot_python_semantic/src/semantic_index/use_def/symbol_state.rs

```diff
@@ -216,9 +262,26 @@ impl SymbolState {
         self.bindings.record_constraint(constraint_id);
     }
 
-    /// Add undeclared as a possibility for this symbol.
-    pub(super) fn set_may_be_undeclared(&mut self) {
-        self.declarations.set_may_be_undeclared();
+    /// Add given visibility constraint to all live bindings.
+    pub(super) fn record_visibility_constraint(
+        &mut self,
+        visibility_constraints: &mut VisibilityConstraints,
+        constraint: ScopedVisibilityConstraintId,
+    ) {
+        self.bindings
+            .record_visibility_constraint(visibility_constraints, constraint);
+        self.declarations
+            .record_visibility_constraint(visibility_constraints, constraint);
+    }
+
+    pub(super) fn reset_visibility_constraints(&mut self, snapshot_state: SymbolState) {
```

Comment:
> Is it the case that if we had perfect simplification of the binary expression trees, this optimization would be unnecessary?

Hm... maybe? Consider this example:
```py
y = 1

if test1:
    pass
elif test2:
    pass

use(y)
```

Ideally, we would see that `y` is always visible. But the constraint that we build is the following:
```
test1 OR (~test1 AND test2) OR (~test1 AND ~test2)
```
If we had binary logic, that condition would always be true. But since we have ternary logic, that is not the case. If both `test1` and `test2` are ambiguous, then the constraint evaluates to ambiguous as well.

The problem is that we also need information about mutual exclusivity. `test1` can be of ambiguous static truthiness, but the pair `(test1, ~test1)` can be either `(true, false)` or `(false, true)`.

So if we somehow go outside the bounds of ternary logic and add this information (maybe using the `is_positive` flag), it might be possible. But staying within the realms of ternary logic, we can't even simplify `test1 OR ~test` to always-true. "Kleene logic has no tautologies".

Reply:
Ok, so I spent some more time thinking about this and reading up on boolean logic, and I think I pointed us in the wrong direction by suggesting three-value logic for this problem. Or not necessarily the _wrong_ direction, as the problem can be solved that way (and you've done that), but not the optimal direction.

This problem is better solved by a regular binary decision diagram, where we first build the full BDD, then in type inference we [restrict](https://docs.rs/oxidd/latest/oxidd/trait.BooleanFunctionQuant.html#method.restrict) it by providing values for the variables (test conditions) that have static truthiness, and then visibility is determined by [satisfiability](https://docs.rs/oxidd/latest/oxidd/trait.BooleanFunction.html#method.satisfiable) of the restricted function (where the only remaining variables in the restricted function would be the test conditions that have ambiguous truthiness). Building a reduced ordered BDD will automatically perform the kinds of simplifications we've discussed here; e.g. I've verified that using OxiDD, building a BDD for the function `test1 OR (~test1 AND test2) OR (~test1 AND ~test2)` does result in simply TRUE. (There are several BDD libraries for Rust; not sure if OxiDD is the best, it's just the one I've been playing with.)

I think we should still go ahead and land this and consider BDD as a further optimization exploration, as we discussed today.